node/heartbeat/monitor.go

```diff
@@ -243,31 +245,33 @@ func (m *Monitor) computeShardID(pubkey string) uint32 {
 }
 
 func (m *Monitor) computeAllHeartbeatMessages() {
+	m.mutHeartbeatMessages.Lock()
```

Comment:
I do not like nested mutex operations. They may lead to deadlocks. 
You can use updateMutex inside heartbeatMessageInfo instances (can move the lock/unlock operations there)

Reply:
+1

Reply:
heartbeatMessageInfo.updateMutex should be used exclusively inside heartbeatMessageInfo struct. The monitor struct that uses heartbeatMessageInfo instances will need to know too much to protect some operations by a mutex.

Reply:
Fixed